### ISSUE TITLE
Disabling root certificate loading should not break TLS

### ISSUE THREAD

**Issue by oakad** [grpc/grpc#27896]
In some (admittedly rare) cases, application will manage it's own TLS security state, starting with unverified TLS and then upgrading the connection (using a combo of `set_server_authorization_check_config` and `set_certificate_provider`). It may be desirable in such apps to not load any root certs at all, especially considering that root cert loading feature only works on sufficiently Linux-like systems.

Unfortunately, right now setting the `GRPC_NOT_USE_SYSTEM_SSL_ROOTS=1` env var results in following:
>E1102 04:50:29.688451234 2511578 ssl_utils.cc:592]           load_file: {"created":"@1635789029.688383147","description":"Failed to load file","file":"external/com_github_grpc_grpc/src/core/lib/iomgr/load_file.cc","file_line":71,"filename":"/usr/share/grpc/roots.pem","referenced_errors":[{"created":"@1635789029.688378415","description":"No such file or directory","errno":2,"file":"external/com_github_grpc_grpc/src/core/lib/iomgr/load_file.cc","file_line":45,"os_error":"No such file or directory","syscall":"fopen"}]}
>E1102 04:50:29.688573953 2511578 ssl_utils.cc:429]           Could not get default pem root certs.
>E1102 04:50:29.688595498 2511578 tls_security_connector.cc:348] Update handshaker factory failed.

And any request is failed with error 14 (failed to connect).

To my understating, this should not happen. When `GRPC_NOT_USE_SYSTEM_SSL_ROOTS` is specified client should behave as if no root certs where present on the system, while still attempting TLS connection as normal (and trusting the application to figure it all out using whatever means).

**Comment by ZhenLian** [comment]
Thanks for opening the issue.

> In some (admittedly rare) cases, application will manage it's own TLS security state, starting with unverified TLS and then upgrading the connection

I don't fully understand the reason for using `unverified TLS`. If you don't want the authentication, what's the point of using TLS at the first place? As far as I am concerned, there is no current way in gRPC to start with "unverified TLS" then upgrade later. 

> It may be desirable in such apps to not load any root certs at all, especially considering that root cert loading feature only works on sufficiently Linux-like systems.

Root certs can work on non-Linux systems. On those systems, you need to specify the location for the root certs through an option object. The thing doesn't work on non-Linux systems is the system-default root location, but you don't need to always store the system root into that default location.

>  When GRPC_NOT_USE_SYSTEM_SSL_ROOTS is specified client should behave as if no root certs where present on the system, while still attempting TLS connection as normal (and trusting the application to figure it all out using whatever means).

Does this happen on the server side or the client side? If on the client side, the root certs are mandatory. If you are setting TLS, the client needs to provide some "evidence" showing which endpoint it trusts, otherwise it is not TLS. `GRPC_NOT_USE_SYSTEM_SSL_ROOTS` is simply a flag helping the system decide whether a system default root is used. If set it to true, you probably want to interpret it as "please don't use the system-default root certs, but use my own root certs", instead of "please don't use root certs at all".

**Comment by oakad** [comment]
I'm not using "unverified" TLS. I'm using custom `set_server_authorization_check_config` and `set_certificate_provider` option classes. Those work outside of default cert set and should keep working even if nothing is loaded from disk.

**Comment by oakad** [comment]
This sort of stuff:
```
       ::grpc::experimental::TlsChannelCredentialsOptions authOptions;
	authOptions.set_server_verification_option(
		GRPC_TLS_SKIP_ALL_SERVER_VERIFICATION
	);

	m_AuthCertChecker = std::make_shared<
		::grpc::experimental::TlsServerAuthorizationCheckConfig
	>(std::make_shared<detail::AllowProperCert>());

	authOptions.set_server_authorization_check_config(m_AuthCertChecker);
	authOptions.set_certificate_provider(
		std::make_shared<detail::MtlsCertificateProviderFactory>(*this)
	);
	authOptions.watch_root_certs();
	authOptions.watch_identity_key_cert_pairs();

        auto ch = ::grpc::CreateChannel(url, ::grpc::experimental::TlsCredentials(authOptions));
```

**Comment by ZhenLian** [comment]
I think our certificate provider doesn't support custom implementations. Is it possible to share some of the definition of `detail::MtlsCertificateProviderFactory`? If using `watch_root_certs`, it will needs to supply the root certs when the system needs them, but that's completely implementation details so I am a bit curious how you achieved that.
Is your application still having problems when you feed your root certs from `detail::MtlsCertificateProviderFactory`? If so, we can take a closer look at the problem, but I guess that's a different problem with this "system-default root certs".

**Comment by oakad** [comment]
My provider works as expected.

1. Let's say there are no certs on disk and the "GRPC_NOT_USE_SYSTEM_SSL_ROOTS" is not set - everything works.
2. Let's say there are certs on disk and the "GRPC_NOT_USE_SYSTEM_SSL_ROOTS" is not set - everything still works for me, but the app allocates half a meg of "reachable on exit" heap blocks for the loaded, but unused certs. No big deal.
3. Let's say there are certs on disk and the "GRPC_NOT_USE_SYSTEM_SSL_ROOTS" is set - the problem from the OP post happens. Why? It sort of expected that it should do the same as p.1.

The point is, this issue is about consistency of behavior during testing.

**Comment by ZhenLian** [comment]
OK I get it. It seems like the problem is, when "GRPC_NOT_USE_SYSTEM_SSL_ROOTS" is set, we shouldn't try to load the default certificates, while in reality that loading code path is somehow triggered. 
The verification logic has changed a lot since the last update. Would you mind trying out the latest HEAD of gRPC and see if the problem still exists(or wait for the next gRPC release)?

**Comment by oakad** [comment]
As with the other issue, I rather hoped to stick to "numbered" releases. But I will try to do some testing with HEAD.

**Comment by erm-g** [comment]
@oakad since few more releases were published, do you still experience the issue?

**Comment by matthewstevenson88** [comment]
I think this is also going to be fixed by #34859.

**Comment by matthewstevenson88** [comment]
Closing because I believe this has been fixed by #34859.
